instrumentation/aws-lambda-1.0/library/src/main/java/io/opentelemetry/instrumentation/awslambda/v1_0/TracingRequestHandler.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.awslambda.v1_0;
+
+import com.amazonaws.services.lambda.runtime.Context;
+import com.amazonaws.services.lambda.runtime.RequestHandler;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Tracer;
+import io.opentelemetry.trace.TracingContextUtils;
+
+/**
+ * A base class similar to {@link RequestHandler} but will automatically trace invocations of {@link
+ * #doHandleRequest(Object, Context)}.
+ */
+public abstract class TracingRequestHandler<I, O> implements RequestHandler<I, O> {
+
+  private final AwsLambdaTracer tracer;
+
+  /** Creates a new {@link TracingRequestHandler} which traces using the default {@link Tracer}. */
+  protected TracingRequestHandler() {
+    this.tracer = new AwsLambdaTracer();
+  }
+
+  /**
+   * Creates a new {@link TracingRequestHandler} which traces using the specified {@link Tracer}.
+   */
+  protected TracingRequestHandler(Tracer tracer) {
+    this.tracer = new AwsLambdaTracer(tracer);
+  }
+
+  /**
+   * Creates a new {@link TracingRequestHandler} which traces using the specified {@link
+   * AwsLambdaTracer}.
+   */
+  protected TracingRequestHandler(AwsLambdaTracer tracer) {
+    this.tracer = tracer;
+  }
+
+  @Override
+  public final O handleRequest(I input, Context context) {
+    Span span = tracer.startSpan(context);
+    Throwable error = null;
+    try (Scope unused = TracingContextUtils.currentContextWith(span)) {
+      return doHandleRequest(input, context);
+    } catch (Throwable t) {
+      error = t;
```

Comment:
👍

i wish there were better way to do this

in other places we are doing the simpler, but less correct 🙁

```
      try (Scope ignored = TRACER.startScope(span)) {
        ...
        TRACER.end(span, response);
        return response;
      } catch (Throwable t) {
        TRACER.endExceptionally(span, t);
        throw t;
      }
```

Reply:
Why is it less correct?

Reply:
it's only less correct if there's an internal exception inside `TRACER.end(span, response)` (at which point maybe we don't care?)

fwiw, the way it's written in this PR maps to how our `@OnMethodEnter`/`@OnMethodExit` behave, which I think makes it nice, if we're going to fail, we fail in a more consistent way

Reply:
Hmm, thinking more on this I think both patterns are wrong. @trask's example seems especially wrong since we shouldn't have one codepath that ends a span with the scope active and another that doesn't since it's inconsistent. Let's imagine `SimpleSpanProcessor`, it may execute an RPC inline with the call to `end`, but will have a different parent for the two cases.

So the question becomes whether it's correct to have the parent in both cases or none of the cases. I guess having the parent is better - `SimpleSpanProcessor` is effectively running span processing, maybe even export, in the context of the request, possibly slowing it down a lot, so it should be traced together with the request.

I'll leave my code as is though since I don't think we have that behavior anywhere in our codebase right now :)

Reply:
@trask Yeah the exception problem means we're breaking user code anyways so much bigger problem :) I'm less concerned about the exception handling though ^^

Reply:
Well, I think I realized that can't work, if `SimpleSpanProcessor` was allowed to emit spans, it would emit as an infinite loop inline with the request :O